File: leetcode/stack-and-queues/0739_daily_temperatures.py
# LC 739. Daily Temperatures | Medium
class Solution:
    # Monotonic stack - O(n)
    def dailyTemperatures(self, temperatures):
        res = [0] * len(temperatures)
        stack = []
        for i, t in enumerate(temperatures):
            while stack and temperatures[stack[-1]] < t:
                j = stack.pop(); res[j] = i - j
            stack.append(i)
        return res

File: leetcode/stack-and-queues/0739_daily_temperatures.cpp
// LC 739. Daily Temperatures | Medium
#include <vector>
#include <stack>
using namespace std;

class Solution {
public:
    vector<int> dailyTemperatures(vector<int>& temp) {
        int n = temp.size();
        vector<int> res(n, 0);
        stack<int> st;
        for (int i = 0; i < n; i++) {
            while (!st.empty() && temp[st.top()] < temp[i]) { res[st.top()] = i - st.top(); st.pop(); }
            st.push(i);
        }
        return res;
    }
};
